islandora.module

```diff
@@ -1685,8 +1690,12 @@ function islandora_object_access($op, $object, $user = NULL) {
  */
 function islandora_islandora_object_access($op, $object, $user) {
   module_load_include('inc', 'islandora', 'includes/utilities');
-
-  return islandora_namespace_accessible($object->id) && user_access($op, $user);
+  if ($object->state != 'A') {
+    return islandora_namespace_accessible($object->id) && user_access($op, $user) && user_access(ISLANDORA_ACCESS_INACTIVE_AND_DELETED_OBJECTS, $user);
```

Comment:
To preserve behavior, and to not resort to dynamically adding permissions to roles, we could do:

-New toggleable feature that can be enabled from admin/islandora (off by default)
-New permission as noted in this PR (`ISLANDORA_ACCESS_INACTIVE_AND_DELETED_OBJECTS`) that is available conditionally given the above feature is enabled. On save `cache_clear_all` to ensure that this permission is available. Object access looks for the feature being enabled before executing the permissions check.

Given sufficient documentation about this feature and its effect, I believe this satisfies A, B, C in the noted use cases. This preserves behavior without the need for an update hook or a hook to dynamically adds permissions on saves.